### ISSUE TITLE
[RML]_MV.R

### ISSUE THREAD

**Issue by takytaky** [takytaky/rusita#3]
# 작업디렉토리 설정

getwd()
setwd("~/[003]_MV")
getwd()

# Model Capacity 그래프 ####

DATA <- read.csv("data/TestData.csv")

str(DATA)
plot(DATA, cex = 0.5)

# lm() 1차함수

Model_1 <- lm(formula = outputs ~ inputs, data = DATA)
Model_1

plot(DATA, pch = 1, cex = 0.5)
abline(Model_1, lwd = 3, col = "red")

# lm() 2차함수

Model_2 <- lm(formula = outputs ~ poly(inputs, 2), data = DATA)
Model_2

a <- seq(0, 1, length = 1000)
points(a, predict(Model_2, data.frame(inputs = a)),
       type = "l", lwd = 3, col = "blue")

# lm() 5차함수

Model_5 <- lm(formula = outputs ~ poly(inputs, 5), data = DATA)
Model_5

points(a, predict(Model_5, data.frame(inputs = a)),
       type = "l", lwd = 3, col = "green")

# lm() 9차함수

Model_9 <- lm(formula = outputs ~ poly(inputs, 9), data = DATA)
Model_9

points(a, predict(Model_9, data.frame(inputs = a)),
       type = "l", lwd = 3, col = "yellow")

# 비교 그래프 작성

par(mfrow = c(2,2))

plot(DATA, pch = 1, cex = 0.5)
abline(Model_1, lwd = 3, col = "red")

plot(DATA, pch = 1, cex = 0.5)
a <- seq(0, 1, length = 1000)
points(a, predict(Model_2, data.frame(inputs = a)),
       type = "l", lwd = 3, col = "blue")

plot(DATA, pch = 1, cex = 0.5)
points(a, predict(Model_5, data.frame(inputs = a)),
       type = "l", lwd = 3, col = "green")

plot(DATA, pch = 1, cex = 0.5)
points(a, predict(Model_9, data.frame(inputs = a)),
       type = "l", lwd = 3, col = "yellow")

par(mfrow = c(1,1))

# Training Error ####

Elec <- read.csv("data/Electric.csv")
str(Elec)

plot(Elec$electricity ~ Elec$surface_area, cex = 1.5)

# 선형모델(1차함수)

Model_1 <- lm(formula = electricity ~ surface_area, data = Elec)
Model_1

# 선형모델(1차함수) Capacity 부족

abline(Model_1, col = "red", lwd = 3)

# 통계요약치

summary(Model_1)

# 실제 전기사용량(실제값)

Real_Elec <- Elec$electricity
head(Real_Elec, 50)

# 선형모델(1차함수) 적용(예측값)

Predicted_Elec_1 <- predict(Model_1, Elec)
head(Predicted_Elec_1, 50)

# Training Error - 선형모델(1차함수) 
# mean((실제값 - 예측값)^2)
# 모든 오차제곱의 평균구하기

Train_Err_1 <- mean((Real_Elec - Predicted_Elec_1)^2)
Train_Err_1

# 비선형모델(2차함수)

Model_2 <- lm(formula = electricity ~ poly(surface_area, 2), data = Elec)
Model_2

# 비선형모델(2차함수) 적용(예측값)

Predicted_Elec_2 <- predict(Model_2, Elec)
head(Predicted_Elec_2, 50)

# 실제값 vs. 선형모델(예측값) vs. 비선형모델(예측값)

Real_Elec[1] ; Predicted_Elec_1[[1]] ; Predicted_Elec_2[[1]]

# Training Error - 비선형모델(2차함수) 

Train_Err_2 <- mean((Real_Elec - Predicted_Elec_2)^2)

# 선형모델(1차함수) vs. 비선형모델(2차함수)

Train_Err_1 ; Train_Err_2

# 비선형모델(5차함수)

Model_5 <- lm(formula = electricity ~ poly(surface_area, 5), data = Elec)
Model_5

# 비선형모델(5차함수) 적용(예측값)

Predicted_Elec_5 <- predict(Model_5, Elec)
head(Predicted_Elec_5, 50)

# Training Error - 비선형모델(5차함수)

Train_Err_5 <- mean((Real_Elec - Predicted_Elec_5)^2)

# 1차함수 vs. 2차함수 vs. 5차함수

Train_Err_1 ; Train_Err_2 ; Train_Err_5

# 비선형모델(9차함수)

Model_9 <- lm(formula = electricity ~ poly(surface_area, 9), data = Elec)
Model_9

# 비선형모델(9차함수) 적용(예측값)

Predicted_Elec_9 <- predict(Model_9, Elec)
head(Predicted_Elec_9, 50)

# Training Error - 비선형모델(9차함수)

Train_Err_9 <- mean((Real_Elec - Predicted_Elec_9)^2)

# 1차함수 vs. 2차함수 vs. 5차함수 vs. 9차함수

Train_Err_1 ; Train_Err_2 ; Train_Err_5 ; Train_Err_9

# 그래프 출력

plot(Elec$surface_area, Elec$electricity, cex = 2)

# 선형모델(1차함수)

abline(Model_1, lwd = 5, col = "red")

# 비선형모델(2차함수)

a <- seq(500, 820, length = 1000)
points(a, predict(Model_2, data.frame(surface_area = a)),
       type = "l", lwd = 5, col = "blue")

# 비선형모델(5차함수)

points(a, predict(Model_5, data.frame(surface_area = a)), 
       type = "l", lwd = 5, col = "green")

# 비선형모델(9차함수)

points(a, predict(Model_9, data.frame(surface_area = a)), 
       type = "l", lwd = 5, col = "yellow")

# Test Error ####
# Training Data와 Testing Data를 7:3으로 분리
# str(Elec) vs. str(Elec[train,]) vs. str(Elec[-train,])

Elec <- read.csv("data/Electric.csv")

# 매번 동일한 값을 추출하기 위해 설정

set.seed(2018)

# Training Data(70%)

TrainData <- sample(1:nrow(Elec), 0.7 * nrow(Elec))

# Testing Data(30%)

Real_Elec <- Elec$electricity[-TrainData]

# 선형모델(1차함수)
# Training Data로 Model_1 생성

Model_1 <- lm(electricity ~ surface_area, data = Elec[TrainData,])

# Testing Data에 선형모델(1차함수) 적용(예측값)

Predicted_Elec_1 <- predict(Model_1, Elec[-TrainData,])

# Testing Error - 선형모델(1차함수)

Test_Err_1 <- mean((Real_Elec - Predicted_Elec_1)^2)
Test_Err_1

# 비선형모델(5차함수)
# Training Data로 Model_5 생성

Model_5 <- lm(electricity ~ poly(surface_area, 5), data = Elec[TrainData,])

# Testing Data에 비선형모델(5차함수) 적용(예측값)

Predicted_Elec_5 <- predict(Model_5, Elec[-TrainData,])

# Testing Error - 비선형모델(5차함수)

Test_Err_5 <- mean((Real_Elec - Predicted_Elec_5)^2)
Test_Err_5

# 비선형모델(9차함수)
# Training Data로 Model_9 생성

Model_9 <- lm(electricity ~ poly(surface_area, 9), data = Elec[TrainData,])

# Testing Data에 비선형모델(9차함수) 적용(예측값)

Predicted_Elec_9 <- predict(Model_9, Elec[-TrainData,])

# Testing Error - 비선형모델(9차함수)

Test_Err_9 <- mean((Real_Elec - Predicted_Elec_9)^2)
Test_Err_9

# Test Error 값이 가장 작은 모델을 선택
# 하지만 Testing Data도 평가과정에서 오염 발생
# 따라서 Testing Data를 Validation Data와 Testing Data로 분리하여 사용
# Model_1 vs. Model_5 vs. Model_9

Test_Err_1 ; Test_Err_5 ; Test_Err_9

# Validation Error ####
# Training Data : Validation Data : Testing Data

CARS <- read.csv("data/Cars.csv")
plot(CARS$mpg ~ CARS$weight, cex = 1.5)

# Training Data : Validation Data : Testing Data(6:2:2) 데이터 분할

set.seed(8888)

# Training Data(60%)

TrainData <- sample(1:nrow(CARS), 0.6 * nrow(CARS))

# Validation Data(20%)

ValidData <- sample(setdiff(1:nrow(CARS), TrainData), 0.2 * nrow(CARS))

# Testing Set(20%)

TestData <- setdiff(1:nrow(CARS), c(TrainData, ValidData))

# 선형모델(1차함수)

Model_1 <- lm(mpg ~ weight, data = CARS[TrainData,])
Pred_W_1 <- predict(Model_1, CARS[ValidData,])
Valid_Err_1 <- mean((CARS$mpg[ValidData] - Pred_W_1)^2)

# 그래프

plot(CARS$mpg ~ CARS$weight, cex = 1.5)
abline(Model_1, lwd = 5, col = "red")

# 비선형모델(9차함수)

Model_2 <- lm(mpg ~ poly(weight, 9), data = CARS[TrainData,])
Pred_W_2 <- predict(Model_2, CARS[ValidData,])
Valid_Err_2 <- mean((CARS$mpg[ValidData] - Pred_W_2)^2)

# 그래프

a <- seq(1800, 5000, length = 10000)
points(a, predict(Model_2, data.frame(weight = a)), 
       type = "l", lwd = 5, col = "blue")

# 선형모델(다중회귀) : horsepower(마력) 추가

Model_3 <- lm(mpg ~ weight + horsepower, data = CARS[TrainData,])
Pred_W_3 <- predict(Model_3, CARS[ValidData,])
Valid_Err_3 <- mean((CARS$mpg[ValidData] - Pred_W_3)^2)

# 평가결과 비교
# 단일회귀 vs. 9차 함수 vs. 다중회귀

Valid_Err_1 ; Valid_Err_2 ; Valid_Err_3

# 개별실습자료 ####

Wage <- read.csv("data/Wage.csv")
str(Wage)
plot(Wage$wage ~ Wage$age)

# Cross Validation ####
# iris(5 x 150)
# Sepal Length(꽃받침 세로길이), Sepal Width(꽃받침 가로길이)
# Petal Length(꽃잎의 세로길이), Petal Width(꽃잎의 가로길이)
# setosa, versicolor, virginica(꽃 종류)

# Linear Model
# 꽃받침 가로길이, 꽃입 세로길이, 꽃입 가로길이 입력
# 꽃받침 세로길이를 예측 및 검증

# install.packages("plyr")
# install.packages("dplyr")

library(plyr)
library(dplyr)

data <- iris

str(data)

# cross validation
# K-fold(k = 5)

k = 5

# 1~5 값을 지정하여 6번째열(id) 생성

data$id <- sample(1:k, nrow(data), replace = TRUE)

# Training과 Validation 구분 벡터 

fold <- 1:k

# 빈 데이터프레임 생성

Prediction <- data.frame()
Test_Result <- data.frame()

# Progress Bar 생성

progress.bar <- create_progress_bar("text")
progress.bar$init(k)

# k = 5인 CV 실행

for(i in 1:k){
  
  # 데이터를 5등분 한 후
  # Train_Data와 Valid_Data 지정
  Train_Data <- subset(data, 
                       id %in% fold[-i])
  
  Valid_Data <- subset(data, 
                       id %in% c(i))
  
  # Train_Data를 사용하여 다중회귀분석 모델 생성
  model_lm <- lm(Train_Data$Sepal.Length ~ .,
                 data = Train_Data)
  
  # 생성된 모델을 Valid_Data에 적용하여 Prediction 값 생성
  temp <- as.data.frame(predict(model_lm, 
                                Valid_Data[, -1]))
  
  # Prediction 데이터프레임에 새로운 값 추가
  Prediction <- rbind(Prediction, 
                      temp)
  
  # Test_Result 데이터프레임에 새로운 값 추가
  Test_Result <- rbind(Test_Result, 
                       as.data.frame(Valid_Data[, 1]))
  
  progress.bar$step()
}

# 실제값과 예측값의 오차계산

result <- cbind(Prediction, Test_Result[, 1])

names(result) <- c("Predicted", "Actual")

result$Difference <- abs(result$Actual - result$Predicted)

# 오차확인

summary(result$Difference)

# CV with Packages ####
# 01 - Validation Set Approach
# 02 - LOOCV(Leave-One-Out Cross-Validation)
# 03 - K-fold Cross-Validaton

CARS <- read.csv("data/Cars.csv")
plot(CARS$mpg ~ CARS$weight, cex = 1.5)

# 01 - Validation Set Approach

set.seed(7088)

# 60% 데이터를 Taining Data로 샘플링

Train_ID <- sample(nrow(CARS):round((nrow(CARS)*0.6)))

# 선형1차 모델

Model_CV_1 <- lm(mpg ~ horsepower, 
                 data = CARS, 
                 subset = Train_ID)
Model_CV_1

mean((CARS$mpg[-Train_ID] - predict(Model_CV_1, CARS)[-Train_ID])^2)

# 비선형3차 모델

Model_CV_3 <- lm(mpg ~ poly(horsepower, 3), 
                 data = CARS, 
                 subset = Train_ID)
Model_CV_3

mean((CARS$mpg[-Train_ID] - predict(Model_CV_3, CARS)[-Train_ID])^2)

# 비선형7차 모델

Model_CV_7 <- lm(mpg ~ poly(horsepower, 7), 
                 data = CARS, 
                 subset = Train_ID)
Model_CV_7

mean((CARS$mpg[-Train_ID] - predict(Model_CV_7, CARS)[-Train_ID])^2)

# 02 - LOOCV(Leave-One-Out Cross-Validation)
# Data Point의 개수만큼 Model을 생성
# Model을 생성 시 한개의 Teat Data를 제외
# 전체 결과의 평균을 적용

CARS <- read.csv("data/Cars.csv")
plot(CARS$mpg ~ CARS$weight, cex = 1.5)

# glm() : Generalized Linear Models
# 선형1차 모델

Model_LOOCV <- glm(mpg ~ horsepower, data = CARS)
Model_LOOCV

# cv.glm() 적용을 위하여 호출

library(boot)

# glm()으로 생성한 모델에 CV를 적용
# cv.glm()
# K argument 생략 시 Data Point의 개수로 적용

LOOCV_ERR <- cv.glm(CARS, Model_LOOCV)

# Cross-Validaton 결과 확인
# 1st : prediction Error
# 2nd : adjusted Cross-Validation estimate(bias)

LOOCV_ERR$delta

# 1차부터 5차까지 Model 비교

# LOOCV() 함수 생성

LOOCV <- function(fit){
  h <- lm.influence(fit)$h
  Err <- mean((residuals(fit) / (1 - h))^2)
  
  return(Err)
}

# LOOCV() 동작 확인

LOOCV(Model_LOOCV)

# 결과 Vector 생성

CV_ERRORs <- rep(0, 5)

# 1차부터 5차까지 값 지정

Degree <- 1:5

# 1차부터 5차까지 모델적용

for(d in Degree){
  Model_LCV <- glm(mpg ~ poly(horsepower, d), data = CARS)
  CV_ERRORs[d] <- LOOCV(Model_LCV)
}

CV_ERRORs

# 결과 시각화

plot(Degree, CV_ERRORs, type = "b", col = "red")

# 03 - K-fold Cross-Validaton

library(boot)

Degree <- 1:5

# 5-fold Cross-Validation

CV_ERRORs_5 <- rep(0, 5)

for(d in Degree){
  Model_LCV <- glm(mpg ~ poly(horsepower, d), data = CARS)
  CV_ERRORs_5[d] <- cv.glm(CARS, Model_LCV, K = 5)$delta[1]
}

CV_ERRORs_5

# 10-fold Cross-Validation

CV_ERRORs_10 <- rep(0, 5)

for(d in Degree){
  Model_LCV <- glm(mpg ~ poly(horsepower, d), data = CARS)
  CV_ERRORs_10[d] <- cv.glm(CARS, Model_LCV, K = 10)$delta[1]
}

CV_ERRORs_10

# LOOCV vs. 5-fold CV vs. 10-fold CV
# 결과 시각화 비교

plot(Degree, CV_ERRORs, 
     type = "b", col = "blue", ylab = "CV Errors")

lines(Degree, CV_ERRORs_5, type = "b", col = "red")
lines(Degree, CV_ERRORs_10, type = "b")

legend("topright", pch = 1,
       c("LOOCV", "5-fold CV", "10-fold CV"),
       col = c("blue", "red", "black"))

# Bias-Variance Tradeoff ####

# Sample Data 생성

OutPut <- c()
InPut <- seq(-10, 10, by = 0.02)

set.seed(1213)

Noise <- rnorm(length(InPut), sd = 0.3)
OutPut <- sin(InPut) + Noise

# Sample Data 시각화

plot(InPut, OutPut, main = "Sample Data")
lines(InPut, sin(InPut), col = "tomato", lwd = 5)

# plots 창 분할(3 x 3)

par(mfrow = c(3, 3))

# 1차부터 9차까지 모델생성 및 시각화

for(i in 1:9){
  Model_T1 <- lm(OutPut ~ poly(InPut, i))
  New_Data <- data.frame(x = InPut)
  Pred <- predict.lm(Model_T1, newdata = New_Data)
  Plot_Name <- paste0(i, " Degree")
  plot(InPut, OutPut, main = Plot_Name, cex = 0.6)
  lines(InPut, Pred, col = "forestgreen", lwd = 3)
}

# 차수에 따른 예측값 list()

Pred_Degree <- list()  

# 차수 별 7개의 Test_Data 구성

for(i in 1:9){
  Pred_Model_Each <- list()
  for(j in 1:7){
    # Boostrap
    T_Data <- data.frame(f = OutPut, x = InPut)
    Test_INDEX <- sample(1:length(InPut), size = 7, replace = FALSE) 
    Train_Data <- T_Data[-Test_INDEX, ]
    Test_Data <- T_Data[Test_INDEX, ]
    Model_T2 <- lm(f ~ poly(x, i), data = Train_Data)
    Pred_Model_Each[[j]] <- data.frame(x = Test_Data$x,
                                       predict.lm(Model_T2,
                                                  newdata = Test_Data))
  }
  Pred_Degree[[i]] <- Pred_Model_Each
}

# Model Variance plot

for(j in 1:9){
  Plot_Name <- paste0(j, " Degree")
  plot(InPut, OutPut, 
       xlim = c(-10, 10), 
       ylim = c(-1.8, 1.8), 
       main = Plot_Name, cex = 0.6)
  for(i in 1:7){
    lines(Pred_Degree[[j]][[i]][ ,1], 
          Pred_Degree[[j]][[i]][ ,2], 
          col = rainbow(6)[i], 
          lwd = 3)
  }
}

par(mfrow = c(1, 1))

for(j in 1:9){
  Plot_Name <- paste0(j, " Degree")
  plot(InPut, OutPut, 
       xlim = c(-10, 10), 
       ylim = c(-1.8, 1.8), 
       main = Plot_Name, cex = 0.6)
  for(i in 1:7){
    lines(Pred_Degree[[j]][[i]][ ,1], 
          Pred_Degree[[j]][[i]][ ,2], 
          col = rainbow(10)[i], 
          lwd = 3)
  }
}
